feat: LLM 서비스 기능 구현

jayn2u opened the pull request:
## LLM 서비스 기능 구현

이 PR은 새길 서버에 LLM(Language Learning Model) 서비스 기능을 추가합니다.

### 주요 기능:
- 텍스트-음성 변환 (ElevenLabs API 사용)
- 음성-텍스트 변환 (OpenAI Whisper API 사용)
- ChatGPT 통합 (텍스트, STT 텍스트, 오디오 URL, 오디오 파일 지원)

### 구현 내용:
- LLM 프록시 컨트롤러 구현
- FastAPI LLM 서버로 요청을 프록시하는 서비스 구현
- 테스트 웹앱 추가 (텍스트-음성, 음성-텍스트, ChatGPT 응답 테스트)

### 테스트:
- 모든 엔드포인트에 대한 테스트 웹앱 구현 완료
- 로컬 환경에서 기능 테스트 완료

jayn2u committed:
chore: tts 패키지로 이관

jayn2u committed:
feat: Whisper를 사용한 transcribe api 설계

- STT로 설계
- GPT에서 응답을 해서 텍스트로 반환

jayn2u committed:
feat: STT -> TTS 엔진 설계 완료

- Spring AI 자체의 불안정성 내포
- LLM Agent의 도움을 많이 받음
- Multipart로 시도해보려고 했지만, 실패

jayn2u committed:
chore: 누락된 클래스 @RequiredArgsConstructor 어노테이션 추가

- 누락된 부분 추가
- 순서 통일시

jayn2u committed:
fix: 어노테이션 개행 및 final 추가

- 어노테이션 개행으로 가독성 개선
- 파라미터의 final 선언자 추가

Co-authored-by: 권예진 <96688810+kwonyj1022@users.noreply.github.com>

jayn2u committed:
fix: 빈줄 제거

Co-authored-by: 권예진 <96688810+kwonyj1022@users.noreply.github.com>

jayn2u committed:
fix: private 함수로 내부 가독성 향상

- 두 단계로 나누어지는 동작을 private 함수로 분리

Co-authored-by: 권예진 <96688810+kwonyj1022@users.noreply.github.com>

jayn2u committed:
fix: 파라미터 개행

- 파라미터 뒤에 괄호를 개행합니다

Co-authored-by: 권예진 <96688810+kwonyj1022@users.noreply.github.com>

jayn2u committed:
fix: 반응형 프로그래밍에서 함수형 호출 가독성 개선

- 체이닝으로 묶이는 코드들을 가독성 좋게 변경

Co-authored-by: 권예진 <96688810+kwonyj1022@users.noreply.github.com>

jayn2u committed:
fix: Multipart 파일 프로세스 코드 제거

jayn2u committed:
fix: 함수 이름 변경에 따른 버그 수정

jayn2u committed:
feat: 사용자 가이드 섹션 추가 및 제목 변경

사용자에게 제공할 안내문을 위한 새로운 intro 섹션을 추가했습니다. 또한, 웹앱 제목을 기존 'Saegil 테스트 웹앱'에서 '새길 서버 테스트 웹앱'으로 변경하여 명확성을 높였습니다.

jayn2u committed:
feat: Swagger 보안 설정 및 주석 추가

각 컨트롤러 메서드에 Swagger 보안 요구사항과 설명 주석을 추가했습니다. 이를 통해 API 문서화와 접근 제어가 강화되었습니다.

jayn2u committed:
docs: API 문서의 설명 및 요약을 한국어로 수정

API 명세의 summary와 description 항목을 한국어로 변경하여 문서의 가독성을 향상시켰습니다. 이를 통해 한국어 사용자를 위한 보다 명확하고 친숙한 문서를 제공합니다.

jayn2u committed:
feat: HTML 스타일 수정으로 가시성 및 읽기 편의성 개선

응답 섹션에 max-height와 스크롤 기능을 추가하여 긴 텍스트 표시를 개선했습니다. 또한 pre 태그에 줄 바꿈 관련 스타일을 추가해 가독성을 향상시켰습니다. 이는 사용자 경험을 위한 UI 개선 작업의 일환입니다.

jayn2u commented:
- `proxy` 패키지 안에 있는 리드미를 작성했는데, 프로젝트에서 제거할까요?

jayn2u commented on src/main/java/com/newworld/saegil/proxy/README_FASTAPI_LLM_SERVER.md:
- `proxy` 패키지 안에 있는 리드미 문서입니다.
- FastAPI 서버의 명세를 작성했습니다.

jayn2u commented on src/main/java/com/newworld/saegil/proxy/README_IMPLEMENTATION.md:
- FastAPI 서버를 연동해서 LLM 프록시 패키지를 만들었습니다.
- SpringBoot 서버 어플리케이션과 함께 사용하는 원리를 작성했습니다.

kwonyj1022 commented on src/main/java/com/newworld/saegil/proxy/config/WebClientConfig.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    private ExchangeFilterFunction logRequest() {
+        return ExchangeFilterFunction.ofRequestProcessor(clientRequest -> {
+            System.out.println("Request: " + clientRequest.method() + " " + clientRequest.url());
```
Sysetm.out.println() 말고 Log 사용해보세요!

kwonyj1022 commented on src/main/java/com/newworld/saegil/proxy/config/WebClientConfig.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    private ExchangeFilterFunction logResponse() {
+        return ExchangeFilterFunction.ofResponseProcessor(clientResponse -> {
+            System.out.println("Response status: " + clientResponse.statusCode());
```
여기도요!

kwonyj1022 commented on src/main/java/com/newworld/saegil/proxy/controller/LlmProxyController.java:
```diff
@@ -0,0 +1,123 @@
+@RequestMapping("/api/llm")
+@RequiredArgsConstructor
+@Tag(name = "LLM 프록시", description = "FastAPI LLM 서버로 요청을 프록시하는 엔드포인트")
+public class LlmProxyController {
```
여기에 있는 api는 다 쓰는 api인가요?

kwonyj1022 commented on src/main/java/com/newworld/saegil/simulator/stt/controller/AudioTranscriptionController.java:
```diff
@@ -0,0 +1,39 @@
+            description = "오디오 파일을 텍스트로 변환합니다",
+            security = @SecurityRequirement(name = SwaggerConfiguration.SERVICE_SECURITY_SCHEME_NAME)
+    )
+    @PostMapping(value = "/transcribe", consumes = MediaType.APPLICATION_OCTET_STREAM_VALUE)
+    public Mono<ResponseEntity<Map<String, String>>> transcribeAudio(
+            @RequestBody byte[] audioFileBytes
+    ) {
+        return speechToTextService.transcribeAndGetAnswer(audioFileBytes)
+                                  .map(ResponseEntity::ok);
+    }
```
transcribeAndGetAnswer 메서드 내부 로직이 실패하면 어떻게 되나요?? 그래도 ok를 응답하나요?

jayn2u replied:
해당 클래스는 이제 제거되었습니다!

kwonyj1022 commented on src/main/java/com/newworld/saegil/simulator/conversation/service/AudioConversationService.java:
```diff
@@ -0,0 +1,43 @@
+    private final TextToSpeechService textToSpeechService;
+
+    // Store the last conversation for retrieval
+    @Getter
+    private Map<String, String> lastConversation = new HashMap<>();
```
copilot이 좋은 피드백을 주었네요! 인증/인가 구현되면 사용자별로 lastConversation 저장하는 쪽으로 수정해야할 것 같아요

jayn2u replied:
대화 목록을 보관하는 로직을 연구해봐야겠습니다. 이건 데이터베이스에 보관을 해야 하는지, API에서 제공하는지도 확인해봐야겠습니다.

kwonyj1022 reviewed (approved):
음 새로운 기술은 어렵군요..

jayn2u committed:
refactor: Spring AI 음성 관련 기능 코드 및 웹앱 제거

음성 대화, 음성 변환, 텍스트-음성 변환과 관련된 컨트롤러, 서비스, HTML 파일 등 모든 관련 코드를 제거했습니다. 더 이상 사용되지 않는 기능으로 판단되어 코드베이스를 정리했습니다.

jayn2u committed:
feat: WebClient 로깅 기능 개선

System.out.println 대신 Logger를 활용하여 요청 및 응답 로깅을 개선했습니다. 이를 통해 로그 관리가 더 효율적이며, 표준화된 로깅 방식으로 전환되었습니다.

jayn2u committed:
refactor: API 경로를 '/api/v1'로 변경

API 버전을 명시하기 위해 모든 엔드포인트의 경로를 '/api/v1'로 업데이트했습니다. 이에 따라 관련된 HTML 파일의 호출 경로도 수정했습니다.

jayn2u commented:
# 업데이트

- `api/v1`로 REST API 주소를 변경했습니다.

jayn2u committed:
docs: 프록시 관련 README 파일 구조와 이름 변경

프록시 패키지 설명서를 개선하고 파일명을 직관적으로 변경했습니다. README_IMPLEMENTATION.md와 README_FASTAPI_LLM_SERVER.md를 각각 proxy-package-guide.md와 fastapi-server-spec.md로 이름 변경하여 명확성을 높였습니다.

jayn2u committed:
docs: LlmProxyController 안드로이드 개발자 가이드 추가

Saegil 서버의 LlmProxyController 사용법을 설명하는 안드로이드 개발자 가이드를 추가했습니다. 이 가이드는 TTS, STT, ChatGPT 연동 구현에 필요한 설정, API 엔드포인트, UI 구성 방법 등을 포함합니다.

kwonyj1022 committed:
refactor: webflux 의존성 제거
